actors/builtin/miner/miner_actor.go

```diff
@@ -444,91 +445,121 @@ func (a Actor) ProveCommitSector(rt Runtime, params *ProveCommitSectorParams) *a
+		)
+		builtin.RequireSuccess(rt, code, "failed to verify deals and get deal weight")
+		AssertNoError(ret.Into(&dealWeights))
+
+		// Request power for activated sector.
+		// Return initial pledge requirement.
+
+		// TODO: since this method gets called by the storage power actor
+		// initially, can we just do this while we're there?
+		// We can probably return the right information from this call to the caller, so it can update there
+		var initialPledge abi.TokenAmount
+		ret, code = rt.Send(
+			builtin.StoragePowerActorAddr,
+			builtin.MethodsPower.OnSectorProveCommit,
+			&power.OnSectorProveCommitParams{
+				Weight: power.SectorStorageWeightDesc{
+					SectorSize:         st.Info.SectorSize,
+					DealWeight:         dealWeights.DealWeight,
+					VerifiedDealWeight: dealWeights.VerifiedDealWeight,
+					Duration:           precommit.Info.Expiration - rt.CurrEpoch(),
+				},
+			},
+			big.Zero(),
+		)
+		builtin.RequireSuccess(rt, code, "failed to notify power actor")
+		AssertNoError(ret.Into(&initialPledge))
 
-		// Verify locked funds are are at least the sum of sector initial pledges.
-		verifyPledgeMeetsInitialRequirements(rt, &st)
+		// Add sector and pledge lock-up to miner state
+		newlyVestedAmount := rt.State().Transaction(&st, func() interface{} {
+			newlyVestedFund, err := st.UnlockVestedFunds(store, rt.CurrEpoch())
+			if err != nil {
+				rt.Abortf(exitcode.ErrIllegalState, "failed to vest new funds: %s", err)
+			}
 
-		// Lock up initial pledge for new sector.
-		availableBalance := st.GetAvailableBalance(rt.CurrentBalance())
-		if availableBalance.LessThan(initialPledge) {
-			rt.Abortf(exitcode.ErrInsufficientFunds, "insufficient funds for initial pledge requirement %s, available: %s", initialPledge, availableBalance)
-		}
-		if err := st.AddLockedFunds(store, rt.CurrEpoch(), initialPledge, &PledgeVestingSpec); err != nil {
-			rt.Abortf(exitcode.ErrIllegalState, "failed to add pledge: %v", err)
-		}
-		st.AssertBalanceInvariants(rt.CurrentBalance())
+			// Unlock deposit for successful proof, make it available for lock-up as initial pledge.
+			st.AddPreCommitDeposit(precommit.PreCommitDeposit.Neg())
 
-		newSectorInfo := &SectorOnChainInfo{
-			Info:               precommit.Info,
-			ActivationEpoch:    rt.CurrEpoch(),
-			DealWeight:         dealWeights.DealWeight,
-			VerifiedDealWeight: dealWeights.VerifiedDealWeight,
-		}
+			// Verify locked funds are are at least the sum of sector initial pledges.
+			verifyPledgeMeetsInitialRequirements(rt, &st)
 
-		if err := st.PutSector(store, newSectorInfo); err != nil {
-			rt.Abortf(exitcode.ErrIllegalState, "failed to prove commit: %v", err)
-		}
+			// Lock up initial pledge for new sector.
+			availableBalance := st.GetAvailableBalance(rt.CurrentBalance())
+			if availableBalance.LessThan(initialPledge) {
+				rt.Abortf(exitcode.ErrInsufficientFunds, "insufficient funds for initial pledge requirement %s, available: %s", initialPledge, availableBalance)
```

Comment:
This can definitely fail, unless we ensure that the precommit deposit to be unlocked covers the initial pledge. As it happens, I think that's exactly what research/cryptoecon will propose for #424. It's complicated by the fact that both those numbers change over time, so the deposit/pledge posted at precommit may no longer cover the pledge when the proof arrives. 

Maybe we can simplify to lock in the calculation at pre-commit. If so, in combination with moving `verifyPledgeMeetsInitialRequirements`, we can probably make this all safe.